core/tests/ledger_cleanup.rs

```diff
@@ -619,23 +589,22 @@ mod tests {
             insert_timer,
             num_shreds_total as f32 / insert_timer.as_s(),
         );
-        let u1 = storage_previous;
 
-        // Poll on some compaction happening
-        info!("Begin polling for compaction ...");
-        let start_poll = Instant::now();
-        while blockstore.storage_size().unwrap_or(0) >= u1 {
-            if start_poll.elapsed().as_secs() > ROCKSDB_FLUSH_GRACE_PERIOD_SECS {
-                break;
-            }
-            std::thread::sleep(Duration::from_millis(200));
+        // Check if LedgerCleanupService has been purging slots to keep the
+        // blockstore under max_ledger_shreds.
+        // LedgerCleanupService won't do anything unless we have had
+        // >=DEFAULT_PURGE_SLOT_INTERVAL slots inserted, so only check if we
+        // have actually inserted that many slots or more.
+        let last_inserted_slot = benchmark_slots - 1;
+        let cleanup_service_may_have_run = last_inserted_slot > DEFAULT_PURGE_SLOT_INTERVAL;
+        if cleanup_service && cleanup_service_may_have_run {
+            let (exceeded_max_ledger_shreds, _, _) = LedgerCleanupService::find_slots_to_clean(
+                &blockstore,
+                last_inserted_slot - DEFAULT_PURGE_SLOT_INTERVAL,
+                max_ledger_shreds,
+            );
+            assert!(!exceeded_max_ledger_shreds);
```

Comment:
So here is the bit we were discussing, a few items:
- With the preset settings (50 slots), we will not enter this if statement
- The logic itself in the if statement is open to hit a race condition for the point you already mentioned of LedgerCleanupService running in the background
- I had to make `LedgerCleanupService::find_slots_to_clean` public in order to use it here to avoid duplicating the logic; also, `DEFAULT_PURGE_SLOT_INTERVAL` arguable shouldn't be public either

I know we said we could decide the fate of this test elsewhere, but I also don't want to spend more time / effort updating a test that we know is more or less obsolete. The test was designed to exercise compaction and could be used to bench insertion rate; compaction is gone and we know the bottleneck is in generation.

Thus, I'm inclined to do one of the following plans:
1) Rip this file out altogether, and add some unit tests in `ledger_cleanup_service.rs` in separate PR
2) Leave the test, but do NOT add this segment I have highlighted. This makes a function public that shouldn't be, and is essentially testing logic that could be tested on `find_slots_to_clean()` directly (hence my above point about unit tests in `LedgerCleanupService`. This would leave this test with no assertions

What do you think @yhchiang-sol ?

Reply:
Great we share thoughts in common.  Here're my thoughts about this:

* Move forward with this PR.  Improvements/changes for the test go to a separate PR.
* This PR can probably remove the ending check of the test.  The test still has value as it can also warn us when the LedgerCleanupService did not clean up in time while the test is running (the soft and hard limit settings).